state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -2266,4 +2314,28 @@ pub(crate) mod utilities {
 			.unwrap_or(u128::MAX),
 		}
 	}
+
+	pub(super) fn split_off_highest_impact_swap<T: Config>(
+		mut swaps: Vec<Swap<T>>,
+		swap_states: &[SwapState<T>],
+		direction: SwapLeg,
+	) -> (Option<Swap<T>>, Vec<Swap<T>>) {
```

Comment:
> mutate the Vec in-place

updated

> In this function, there are some key invariants that we can enforce/assert on

I will add debug asserts for these. However, I'm not sure whether we should change the signature: that would mean asserting in release too, and I thought we aim to avoid runtime panics. It is one thing to assert on something locally obvious, but in this case there is enough "action at a distance" that I'd rather have a fallback way of handling a scenario that we think is impossible now.

Note that if this invariant is violated doesn't (necessarily) mean that something has gone so wrong that panicking is the lesser evil. Far more likely would be that we made a mistake in `do_group_and_swap` (or some other similar code that we will write in the future) where `SwapLegFailed` is incorrectly reported and a totally reasonable/desirable way to handle this would be to simply give up on the entire batch (which is what I'm doing here).